Final definite assignment (#1091)
dart-lang/language: accepted/future-releases/nnbd/feature-specification.md

It is an error to call an expression whose type is potentially nullable and not
`dynamic`.

It is an error if a top level variable or static variable has a non-nullable
type and no initializer expression unless the variable is marked with the `late`
modifier.

It is an error if a class declaration declares an instance variable with a
potentially non-nullable type and no initializer expression, and the class has a
generative constructor where the variable is not initialized via an initializing
formal or an initializer list entry, unless the variable is marked with the
`late` modifier.

It is an error if a mixin declaration or a class declaration with no generative
constructors declares an instance variable with a potentially non-nullable type
and no initializer expression unless the variable is marked with the `late`
modifier.

It is an error to derive a mixin from a class declaration which contains an
instance variable with a potentially non-nullable type and no initializer

> expression unless the variable is marked with the `late` modifier.

Author: For the nth time now.... :)  This is text that @lrhn  added in February, and you reviewed.  I have no context on it whatsoever - I just reformatted it.  You both have called it out multiple times now, so I believe you that there is some issue that should be resolved with this, but again, I have **no context whatsoever** on why this was added and what is missing, so maybe you guys should follow up separately?

It is an error for a class with a `const` constructor to have a `late final`
instance variable.

> It is not a compile time error to write to a `final` non-local variable or field

Author: I'll change it to instance variable.  Just using "non-local variable" to mean field is too confusing to me - fields aren't variables semantically (one way to see this is to note that they don't admit alpha-variance).  "Instance variable" is consequently an odd term, but at least it's unambiguous.

> It is not a compile time error to write to a `final` non-local variable or field

Reviewer: I haven't been worried about the term 'instance variable' (and the language specification is using it 70 times vs. just two occurrences of 'field' in normative text), but it sounds like we'd consider 'instance variable' acceptable for now.

However, I don't understand why this rule would not be applicable to a library variable or a class variable. As far as I can see, 'non-local variable' covers exactly the three cases we need, and 'instance variable' covers only one of them.

It is a warning to use the null check operator (`!`) on an expression of type
`T` if `T` is **strictly non-nullable** .

### Local variables and definite (un)assignment.

As part of the null safety release, errors for local variables are specified to
take into account **definite assignment** and **definite unassignment** (see the
section on Definite Assignment below).  We say that a variable is **potentially
assigned** if it is not **definitely unassigned**, and that a variable is
**potentially unassigned** if it is not **definitely assigned**.

In all cases in this section, errors that are described as occurring on reads of
a variable are intended to apply to all form of reads, including indirectly as

> part of composite assignment operators, as well as via pre and post-fix

Reviewer: The standard phrase for `+=` and such is 'compound' assignment operator.
```suggestion
part of compound assignment operators, as well as via pre and post-fix
```

Author: Done.

operators.  Similarly, errors that are described as occurring on writes of a
variable are intended to apply to all form of writes.

It is a compile time error to assign a value to a `final`, non-`late` local
variable which is **potentially assigned**.  Thus, it is *not* an error to
assign to a **definitely unassigned** `final` local variable.

It is a compile time error to assign a value to a `final`, `late` local variable

> if it is **definitely assigned**. Thus, it is *not* an error to assign to a

Reviewer: It may cause a dynamic error, so maybe we'd spell it out:
```suggestion
if it is **definitely assigned**. Thus, it is *not* a compile-time error to assign to a
```

Author: Done.

| var x;            | Ok            | Ok                  | Ok              |
| final x;          | Error         | Error               | Ok              |
| int x;            | Ok            | Ok                  | Ok              |
| int? x;           | Ok            | Ok                  | Ok              |
| final T x;        | Error         | Error               | Ok              |
| late var x;       | Ok            | Ok                  | Ok              |
| late final x;     | Error         | Ok                  | Ok              |
| late T x;         | Ok            | Ok                  | Ok              |
| late final T x;   | Error         | Ok                  | Ok              |

### Local variables and inference

Local variables with explicitly written types are given the declared types as
written.  The declared type of the variable is considered a "type of interest"
in the sense defined in the flow analysis specification.  If the variable has an
initializer (explicit or implicit) and is not `final`, then the declaration is
treated as an assignment for the purposes of promotion.

*Treating the declared type of the variable as a "type of interest" implies that

> if the variable is a nullable type, then its non-nullable version is also a type

Reviewer 2: Maybe:

"is a nullable type..." -> "has a nullable type, then the type's non-nullable version is also a type"

For what it's worth, in the doc I wrote for null safety, I use "underlying type" to refer to `T` in `T?`. Both "version" and "base" are pretty overloaded words, so that seemed fairly clear to me.

Author: Done.

of interest.  Treating the initialization as an assignment for the purposes of
promotion means that initializing a mutable variable declared at type `T?` with
a value of non-nullable type `T` immediately promotes the variable to the
non-nullable type.*

```dart
void test() {
  int? x = 3; // x is declared at `int?`
  x.isEven; // Valid, x has been promoted to `int`
  x = null; // Valid, demotes to the declared type.
}
```

Local variables with no explicitly written type but with an initializer are

> given an inferred type equal to the type of their initializer, unless the that

Reviewer: ```suggestion
given an inferred type equal to the type of their initializer, unless that
```

Author: Done.

for the variable.  The **initialization inferred type** of the variable is
always treated as a **type of interest** for the purposes of promotion
(including its non-nullable version if applicable).

```dart
void test() {
  var x; // x is an untyped variable
  x = 3; // x is given the provisional initialization inferred type of `int`
  var y; // y is an untyped variable
  y = null; // y is given the provisional initialization inferred type of `Null`
}
```

```dart
void test() {
  // Error, a type cannot be inferred for x
  var x;
  while(b) {
    x = 3; // Error: not an initializing assignment, so no type for `x`
  }
}
```

If de-promotion is performed to an **untyped variable** which has a
**provisional initialization inferred type**, then the **provisional
initialization inferred type** becomes the **final initialization inferred
type** for the variable, the variable is de-promoted to that type.

```dart
void test() {
  var x;
  x = (3 as num); // x is given the provisional initialization inferred type of `num`
  if (x is! int) return;
  while(b) { // x is de-promoted to num, and num is made the final initialization inferred type
    x = 3;
  }
}
```

> On joins with back edges in the flow analysis, **untyped variables** are

Reviewer 3: Note: the flow analysis spec calls these joins "conservative joins" rather than "joins with back edges".  The rationale is that we have one circumstance where we do conservative joins even though there is no back edge: when analyzing `try` statements.  E.g., in the code below, `x` is promoted to `int` at both the top and bottom of the `try` block, but there is some point during the `try` block where it's not promoted, so we need to demote it during processing of the `catch`.  That's implemented using "conservative join" even though there's no back edge involved.
```dart
Object x = ...;
if (x is! int) return; // Promotes to int
try {
  x = ...; // De-promotes
  ...; // Might throw exception
  if (x is! int) return;
  ...; // Might throw exception
} catch (_) {
  x.isEven; // Error: x not known to be `int`
}
```
I think in all the cases in this doc where you talk about "joins with back edges", the same logic will apply to these sorts of try/catch scenarios.  So you might consider rephrasing some of the text in this doc to use "conservative join" terminology rather than talking about back edges.

inferred type** of a variable after a join point is specified as follows.

At a non-backed edge join point in the program, if a non-`late` **untyped
variable** has one or more *reachable* paths into the join on which it has no
**initialization inferred type**, and one or more *reachable* paths into the
join on which it has an **initialization inferred type**, it is an error and
inference fails for the variable. In other words, a non-`late` variable must
either be assigned a value on every reachable path or on no reachable paths.

At a non-back edge join point in the program, if an **untyped variable** has no
**initialization inferred type** on every path into the join, then it has no
**initialization inferred type** after the join.

If neither of the above cases apply, then there is at least one path (reachable
or not) into the join which has an **initialization inferred type**, and the
**initialization inferred type**  for the variable after the join point
is computed as follows:
  - If any path into the join (reachable or not) has a **final initialization
    inferred type** `T`, then:
    - It is an error if any other path (reachable or not) has a **final

>       initialization type** `S` and `S` is not equal to `T`.

Reviewer 3: Is "equal" defined in terms of normalization?  E.g. is `FutureOr<int?>` "equal" to `FutureOr<int?>?` for purposes of this check?

    - It is an error if any other path (reachable or not) has the **provisional
      initialization type** `Null` and `T` is not nullable.
    - It is an error if any other path (reachable or not) has a **provisional
      initialization type** `S` such that `S` is not `Null`, and neither

>       `S` nor `S?` are equal to `T`.

Reviewer 3: Same issue with the definition of "equal" here

    - Otherwise, the **final initialization type** after the join is `T`.
  - Otherwise, if the **provisional initialization inferred type** on each path
    which has one (reachable or not) is the same type `T` (using syntactic
    equality), then the **provisional initialization inferred type** after the
    join is `T`.
  - Otherwise, if the **provisional initialization inferred type** on each path
    which has one (reachable or not) is one of `T`, `T?` or `Null` for some type

>     `T` (using syntactic equality), then the **provisional initialization

Reviewer 3: I'm concerned about using syntactic equality here rather than normalized equality.  Does that mean that if the provisional initialization inferred type on one of the paths is `Never?`, then this rule doesn't apply?  That seems bad.

    if (b) {
      u = 0 as num; // u is promoted to num
      if (u is! int) u = 1;
      // u has inferred type num, but is promoted to int
    } else {
      u = 3;
      // u is inferred as int
    } Error: type cannot be inferred for u
  }
}
```

It is an error to read an **untyped variable** when it has no **initialization
inferred type**.

It is an error to write to an **untyped variable** when it has no
**initialization inferred type**, unless the write is an **initializing
assignment**.

```dart
void test(bool b) {
  {
    var x; // x is an untyped variable
    print(x); // Error, x is used before it is defined
  }

  {
    late var x; // x is an untyped variable
    var f = () => print(x); // Error, type cannot be inferred for x
  }

  {
    var x; // x is an untyped variable
    if (b) {
      print(x); // Error: x has no initialization inferred type.
      x = 3;
    } else {
      x = 4;
    } // x is inferred as type int
    print(x); // No error here, since x has been inferred at this point
  }

  {
    var z;
    do {
      // z is potentially assigned.
      if (b) {
        z = 42;
      } else {
        z = 37;
      }
    } while (something);

>     print(z); // Error, type cannot be inferred for z

Reviewer: The `do` loop is actually a bit more analyzable than a `while` loop because it is guaranteed to run at least once. So if the body considered without the back-edge gives rise to a typing, it should be sound to use that. The fact that it may repeat will still need to be taken into account, of course, such that we'll get an error if the loop body is writing to a `final` variable. But that might just be an extra run of the flow analysis on the loop body in a state where `z` is considered potentially assigned, checking for those "don't reassign" constraints.

Reviewer 3: Personally I would rather not complicate flow analysis further by adding a special rule for `do` loops, especially if that means that we have to run flow analysis twice on the loop body.  It was an explicit design goal of flow analysis to make sure we didn't have to run it twice, since the CFE and analyzer resolution logic that it's fitting into is designed in such a way that doesn't permit running twice.  IMHO, `do` loops are rare enough in practice that the extra complexity we would take on by making this change wouldn't carry enough user benefit to justify it.

  }
```

Notice that the above definitions never discard an **initialization inferred
type** from any path, and always ensure that the **initialization inferred
type** is consistent across joins.  Any variable which is given an
**initialization inferred type** in this way may be treated by IDEs and tools as
having this type for the purposes of documentation and error messages. An
untyped variable which is never read nor written may have no **initialization
inferred type**.  Such a degenerate variable may be treated by the tools as
having type `Object?` for the purposes of documentation and error messages.

```dart
void test(b) {
  var x; // x is an untyped variable, documented as having type int?
  var z; // z is an untyped variable, documented as having type Object?
  if (b) {
    x = 3; // x is inferred to int
    return;
  } else if (b) {
    return;  // Note, x has no inferred type on this path
  } else {
    // x not assigned

>   } // x has type `int?`

Reviewer: Why don't we get an error: `x` is non-`late`, cf. line 862?
